src/batSplay.ml

```diff
@@ -769,6 +794,37 @@ struct
     | Empty -> raise Not_found
     | Node (l, kv, r) -> kv, sref (bst_append l r)
 
+
+  let add_seq s m =
+    BatSeq.fold_left
+      (fun m (k, v) -> add k v m)
+      m
+      s
+    
+  let of_seq s =
+    add_seq s empty
+
+  let rec to_seq m =
+    fun () ->
+    (* since the tree can change shape arbitrarily, fetch all bindings and turn them into a seq. *)
+    BatSeq.of_list (bindings m) ()
```

Comment:
This is not a very nice implementation. The explanation is not correct (a splay tree is a mutable pointer to a persistent tree structure, so you could just dereference the tree and then iterate on it on-demand), but more importantly the work of providing an iterator on splay tree elements has already been done for the `Enum` datatype and it is easy to reuse it. Look at the type `'a enumeration`, the function `cons_enum` and the way it is used in the functions `equal` or `compare` (they traverse two trees, you only need one, but the idea is the same).

Reply:
I did not realize that it is possible to use the persistent tree structure. In hindsight it seems obvious...
I will rewrite this, i will probably copy the implementation from BatSet.to_seq.

Reply:
> i will probably copy the implementation from BatSet.to_seq or BatMap.to_seq.

We designed a datatype (`'a enumeration`) specifically to provide an efficient solution to this problem, and I am not sure why you don't want to use it.

Reply:
So, i tried to implement to_seq with 'a enumeration and without:
```
  let to_seq m () = (* version with cons_enum *)
    let rec to_seq_hlp m =
      match m with
      | End -> BatSeq.Nil
      | More(k, v, r, e) ->
         BatSeq.Cons ((k, v), fun () -> to_seq_hlp (cons_enum r e)) in
    to_seq_hlp (cons_enum (sget m) End)

  let to_seq m = (*  version without cons_enum *)
    let rec to_seq_hlp m =
      fun () ->
      match m with
      | Empty -> BatSeq.Nil
      | Node(l, kv, r) ->
         BatSeq.append (to_seq_hlp l) (fun () -> BatSeq.Cons (kv, to_seq_hlp r)) () in
    to_seq_hlp (sget m)
```

I can see that `cons_enum` and the `'a enumeration` data type might be preferable to generating closures.
However, creating a `'a seq` requires creating one closure per item anyways, so we're not gaining anything.

For `to_seq_from` it gets worse: either we reimplement the `cons_enum` function to compare with the given key and skip descending into the left subtree, or we generat a full enumeration and throw away the leading N elements that are too small.

Reply:
Let me first apologize for the somewhat grumpy tone of my last remark. I will look at using `'a enumeration` in BatSeq at first and see if we have a measurable performance difference, before we decide what to do for BatMap and BatSplay.

Reply:
I will try to reimplement using enumeration.

Reply:
I proposed a benchmark for `BatSet.to_seq` in #1010. It seems that the enumeration/iteration-based implementation is sensibly faster. (I don't find it surprising, but it is better to have hard data than guesses.) I proposed an implementation, still for `BatSet`, in #1011; for `to_seq_from` I indeed used the approach you suggest of a new `cons_iter_from` function. Feel free to comment on the implementations, and/or to reuse the idea for Map/Splay if you find it convincing.

Reply:
After implementing to_seq and to_seq_from for BatSplay i also used the implementation for BatMap and BatSet.
You can see my implementations in the PR, they look remarkably similar to your implementation in the benchmark.